src/core/server/saved_objects/service/lib/repository.ts

```diff
@@ -989,3 +1444,48 @@ function getBulkOperationError(error: { type: string; reason?: string }, type: s
       };
   }
 }
+
+/**
+ * Returns an object with the expected version properties. This facilitates Elasticsearch's Optimistic Concurrency Control.
+ *
+ * @param version Optional version specified by the consumer.
+ * @param document Optional existing document that was obtained in a preflight operation.
+ */
+function getExpectedVersionProperties(version?: string, document?: SavedObjectsRawDoc) {
+  if (version) {
+    return decodeRequestVersion(version);
+  } else if (document) {
+    return {
+      if_seq_no: document._seq_no,
+      if_primary_term: document._primary_term,
+    };
+  }
+  return {};
+}
+
+/**
+ * Returns the string representation of a namespace.
+ */
+function getNamespaceString(namespace?: string) {
+  return namespace ?? 'default';
+}
+
+/**
+ * Returns a string array of namespaces for a given saved object. If the saved object is undefined, the result is an array that contains the
+ * current namespace. Value may be undefined if an existing saved object has no namespaces attribute; this should not happen in normal
+ * operations, but it is possible if the Elasticsearch document is manually modified.
+ *
+ * @param namespace The current namespace.
+ * @param document Optional existing saved object that was obtained in a preflight operation.
+ */
+function getSavedObjectNamespaces(
+  namespace?: string,
+  document?: SavedObjectsRawDoc
+): string[] | undefined {
+  if (document) {
+    return document._source.namespaces;
```

Comment:
I saw `document._source?.namespaces ?? []` in some places (I think?). We don't want it there?

Reply:
We don't want to coalesce to an array here if the document has no `namespaces`, because of the ways that consumers use this function. However, I suppose it wouldn't hurt to include the optional chaining!